### PR TITLE
Unit-Tests

### PR DESCRIPTION
**Pull request by moritzlayer**
Add a unit test suite for all files in lif_meanfield_tools written with pytest.

I took the unit tests for input_output.py incorporated in Simon's pull request #34 and merged them into this one.

### CODE REVIEW

File: tests/fixtures/config/analysis_params_test.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+f_min:
+  val: 0.1
+  unit: Hz
+
+f_max:
+  val: 300
+  unit: Hz
+
+df:
+  # 1/(2*pi)
```
Comment (essink) [inline tests/fixtures/config/analysis_params_test.yaml:12]: This comment can be deleted.

File: tests/unit/test_meanfield_calcs.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+                                                   std_params_eval_spectra):
+        check_exception_is_raised_if_k_is_too_large(self.func,
+                                                    std_params_eval_spectra)
+
+    def test_exception_is_raised_if_prop_is_singular_and_inv_prop_is_requested(
+            self, output_test_fixtures):
+        params = output_test_fixtures.pop('params')
+        params['quantity'] = 'eigvals'
+        params['matrix'] = 'prop_inv'
+        # import pdb; pdb.set_trace()
```
Comment (essink) [inline tests/unit/test_meanfield_calcs.py:241]: Delete debugging remainder.

```suggestion
```

File: pytest.ini [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[pytest]
+norecursedirs = .* lif_meanfield_tools example *.egg dist build
```
Comment (essink) [inline pytest.ini:2]: I think, I don't really understand what it does: Does it tell pytest which directories should not recurse into?

If yes one could add:
`readme_figures`

and maybe change
```suggestion
norecursedirs = .* lif_meanfield_tools example *.egg* dist build
```
to cover `.egg-info`, which exists in my case.

**Review by jhnnsnk — CHANGES_REQUESTED** [review]
Thanks for this extensive test suite.
Before merging, please add some documentation about how to use the tests in the README and add a file specifying the environment (without prefix).

File: tests/fixtures/config/analysis_params_test.yaml (hunk 1 above)
Comment (moritzlayer) [reply tests/fixtures/config/analysis_params_test.yaml:12]: done

File: tests/fixtures/create_fixtures.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
+#!/usr/bin/env python
+# encoding:utf8
+'''
+Creates fixtures for lif_meanfield_tools tests.
+
+WARNING: Only use this script, if your code is trustworthy! The script runs
```
[the remark this answers is not in the document]
Comment (moritzlayer) [reply tests/fixtures/create_fixtures.py:6]: You are right, but then I will keep it for now.

File: tests/unit/test_network.py [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+                          quantity=1 * ureg.s,
+                          string='test_string',
+                          array=np.array([1, 2, 3]),
+                          quantity_array=np.array([1, 2, 3]) * ureg.s,
+                          list_of_quantities=[1 * ureg.s,
+                                              2 * ureg.s,
+                                              3 * ureg.s],
+                          )
+result_ids = sorted(result_types.keys())
+result_ids = sorted(result_types.keys())
```
[the remark this answers is not in the document]
Comment (moritzlayer) [reply tests/unit/test_network.py:331]: Done

File: pytest.ini (hunk 3 above)
Comment (moritzlayer) [reply pytest.ini:2]: Yes, you're right. These are the paths that are excluded during test collection. `.egg-info` is already excluded due to the `.*` expression.  I will add `readme_figures` though.

File: tests/unit/test_input_output.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        filename = param_test_dict['label'] + '_' + hash + '.h5'
+        output_key = 'results'
+        tmp_test = tmpdir.mkdir('tmp_test')
+        with tmp_test.as_cwd():
+            io.save(output_key, param_test_dict, filename)
+            loaded_params = io.load_from_h5(param_test_dict)
+        params = loaded_params[output_key]
+        check_quantity_dicts_are_equal(params, param_test_dict)
+    
+    def test_save_and_load_existing_results_with_anlysis_params(
```
[the remark this answers is not in the document]
Comment (moritzlayer) [reply tests/unit/test_input_output.py:269]: It is amazing that you found that!

File: tests/unit/test_meanfield_calcs.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+    power_spectra,
+    eigen_spectra,
+    additional_rates_for_fixed_input,
+    effective_coupling_strength)
+
+
+class Test_firing_rates:
+    """Tests firing rates function. Probably this is a functional test."""
+
+    # Def function as staticmethod because it is not tightly related to class,
```
[the remark this answers is not in the document]
Comment (moritzlayer) [reply tests/unit/test_meanfield_calcs.py]: Done

File: tests/unit/test_input_output.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+
+class Test_load_h5:
+    
+    @pytest.mark.xfail
+    def test_raise_exception_if_filename_not_existing(self, tmpdir):
+        tmp_test = tmpdir.mkdir('tmp_test')
+        filename = 'test.h5'
+        with tmp_test.as_cwd():
+            with pytest.raises(IOError):
```
[the remark this answers is not in the document]
Comment (moritzlayer) [reply tests/unit/test_input_output.py:238]: Converted into an issue

**Comment by moritzlayer** [comment]
I added some documentation to the `README.md`, an `environment.yaml`, and I integrated Simon's suggestions.

File: README.md [hunk 9]
```diff
@@ -226,6 +226,68 @@ publication in Physical Review Research):
 - __xi_of_k__: Effective spatial profile (see Fig. 3(b) and (d)).
 - __solve_chareq_rate_boxcar__: Analytical solution of the characteristic
   equation for a neural-field model with boxcar-shaped connectivity kernels.
+  
+# Testing
+
+We have an extensive test suite using the `pytest` framework. If you want to
+run all the tests, you can simply do so by installing and activating the conda environment specified in the provided `environmen.yaml` file, and running
```
Comment (jhnnsnk) [inline README.md:233]: Fix environment.yaml with a t.

File: README.md [hunk 10]
```diff
@@ -226,6 +226,68 @@ publication in Physical Review Research):
...
+`create_fixtures.py` using the parameters defined in `config/`.
+
+`unit/` contains all unit tests as well as a file `checks.py` which is a
+collection of custom assert functions.
+
+## Test Design
+
+Many test classes define the tested function as `staticmethod`, because the
+function itself is not tightly related to class, but we still want to attach it
+to the class for later reference. This allows calling function as an 'unbound function', without passing the instance to the function:
```
Comment (jhnnsnk) [inline README.md:275]: "to allow **sb.** to do sth.", also below

File: README.md [hunk 11]
```diff
@@ -226,6 +226,68 @@ publication in Physical Review Research):
...
+```
+pytest
+```
+from the root directory (the one containing `tests` and `lif_meanfield_tools`).
+If you want to be more specific, you can run single tests as well
+```
+pytest tests/unit/test_meanfield_calcs.py::Test_firing_rates::test_correct_output
+```
+See the `pytest` documentation for all available options.
+
```
Comment (jhnnsnk) [inline README.md:243]: I think it would be good to comment on why some tests fail and that failure is unequal to an error.

**Review by jhnnsnk — CHANGES_REQUESTED** [review]
Thanks for addressing my comments. I can confirm that everything runs through with the new environment.
I only have some minor further requests regarding the documentation.

**Review by jhnnsnk — APPROVED** [review]
Thank you, looks all great now! :+1:

**Pull request merged**